### PR TITLE
Use flexx.pyscript to allow Python callbacks

### PR DESCRIPTION
**Pull request by almarklein**
issues: implements part of #2788 
replaces #2799

Done in Bokeh:
* Modifies the CustomJS class to allow using it as a decorator for a callback function.
* Adds an example
* Adds docs (note that the result plot for the example in the docs does not work)

Done in Flexx:
* PyScript is now more Pythonic. Probably enough for most code to work as-is. Users don't need to know about JavaScript's quircks in most cases.
* PyScript now works on all Python versions 3.3+ (including pypy)
* The above is tested with CI
* Released, so it can be pip-installed.

To discuss:
- [x] I propose that flexx is an optional dependency. When trying to use this feature when flexx is not installed, we raise an informative exception,
- [x] Flexx can be pip-installed. Should we make it conda installable?

### CODE REVIEW

**Comment by damianavila** [comment]
>I propose that flexx is an optional dependency. When trying to use this feature when flexx is not installed, we raise an informative exception,

I agree...

>Flexx can be pip-installed. Should we make it conda installable?

Sure... we need to...

**Comment by mattpap** [comment]
> Should we make it conda installable?

Technicalities aside, yes, it must be. In fact, very package that we depend on or use internally must be conda installable.

**Comment by mattpap** [comment]
Perhaps this is a story for another PR, but would it be possible to define custom models like this as well?

**Comment by birdsarah** [comment]
The `awesomeness : lines_of_code` ratio is off the charts!!! 

I know this is because of all the hard work you've been doing on flexx. Congrats. I'm psyched about trying this out.

**Comment by jakirkham** [comment]
This sounds really cool @almarklein. Looking forward to moving from JavaScript in strings to decorated Python functions.

**Comment by almarklein** [comment]
What is the best way to make this available as a conda package? Add a recipe to the conda-recipes repo, or add a recipe to the flexx repo?

> Perhaps this is a story for another PR, but would it be possible to define custom models like this as well?

@mattpap I *think* it would be possible to write classes that subclass `PlotObject`, or any other JS-side class, so yes, I think this is feasible, though better done in a separate PR.

**Comment by almarklein** [comment]
Is there a way to make an example *not* run on the 2.7 exampples_flake_docs?

**Comment by bryevdv** [comment]
> Perhaps this is a story for another PR, but would it be possible to define custom models like this as well?

This thought occurred to me as well. We could imagine shipping some of our models this way, but that's not really practical until there is 2.7 support. That also cuts off some models from other language bindings. So then we might imagine using this to compile at build time rather than runtime, of course that would put us in the weird position of needing python 3 to build python 2 packages.

**Comment by bryevdv** [comment]
@almarklein edit https://github.com/bokeh/bokeh/blob/master/tests/travis/test.yaml

**Comment by almarklein** [comment]
@bryevdv can I edit that in a way to only skip on 2.7? It would be nice to still test the PyScript examples on 3.x. If not, I guess I better add a unit test.

I guess the need for Python 2.7 support is already increasing ...

**Comment by bryevdv** [comment]
The test config is general and  not specific to any Python version afaik.

**Comment by jakirkham** [comment]
Oh, some how I missed that this doesn't work on Python 2. Is there a plan to get this working on Python 2? Also, if so, what sort of timeline are we looking at?

**Comment by almarklein** [comment]
We initially targetted Python 3 only because that's what Flexx is currently targeting, and add 2.7 support when users request it... which you just did :)  Not exactly sure when I'm able to pick this up, but probably within 3 months or so.

**Comment by bryevdv** [comment]
@almarklein @damianavila we need a conda package (on the bokeh channel, at least) for flexx. Nothing else in the test execution uses pip at all anymore I don't want to add a complicating special case just for this,

File: sphinx/source/docs/user_guide/interaction.rst [hunk 1]
```diff
@@ -290,3 +290,26 @@ interactions such as a box zoom, wheel scroll or pan.
...
 
 .. |bokeh.plotting| replace:: :ref:`bokeh.plotting <bokeh.plotting>`
+
+.. _userguide_interaction_actions_in_python:
+
+CustomJS with a Python function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A CustomJS callback can also be implemented as a Python function, which
+is then transpiled to JavaScript using PyScript. This makes it easier
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction.rst]: "transpiled" is a very techy term. Maybe just "translated" or even "turned into"

File: sphinx/source/docs/user_guide/interaction.rst [hunk 2]
```diff
@@ -290,3 +290,26 @@ interactions such as a box zoom, wheel scroll or pan.
...
+for users to define client-side interactions without having to learn
+JavaScript. To use this functionality you need the Flexx library
+(install with ``pip install flexx``). Note that this functionality is
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction.rst]: Use an admonition:
```
.. note::
    This capability is currently only available on Python 3.x.
```

Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction.rst]: Also see note in main conversation about a conda package.

File: sphinx/source/docs/user_guide/interaction.rst [hunk 3]
```diff
@@ -290,3 +290,26 @@ interactions such as a box zoom, wheel scroll or pan.
...
+currently only supported on Python 3.x.
+
+For more information about the subset of Python that is supported in
+callbacks, see the `PyScript documentation
+<http://flexx.readthedocs.org/en/latest/pyscript/index.html>`_.
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction.rst]: Use a "footnote" style reference. See, e.g., all the references here: https://raw.githubusercontent.com/bokeh/bokeh/master/sphinx/source/index.rst

File: sphinx/source/docs/user_guide/source_examples/interaction_callbacks_in_python.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+x = [x*0.005 for x in range(0, 200)]
+y = x
+
+source = ColumnDataSource(data=dict(x=x, y=y))
+
+plot = figure(plot_width=400, plot_height=400)
+plot.line('x', 'y', source=source, line_width=3, line_alpha=0.6)
+
+@CustomJS
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/source_examples/interaction_callbacks_in_python.py]: Hmm, I understand this but I'm not sure I like this exactly as-is. It's a little surprising to see a class used as a decorator, and also CamelCase for decorators is jarring as well. It's also not "wrapping" the function and returning a new function, which is the typical usage people would expect. I would say show explicit usage by created the a `CustomJS` and passing it `callback` as an argument.

File: examples/plotting/file/slider_pycalback.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+x = np.linspace(0, 10, 500)
+y = np.sin(x)
+
+source = ColumnDataSource(data=dict(x=x, y=y))
+
+plot = figure(y_range=(-10, 10), plot_width=400, plot_height=400)
+plot.line('x', 'y', source=source, line_width=3, line_alpha=0.6)
+
+@CustomJS
```
Comment (bryevdv) [inline examples/plotting/file/slider_pycalback.py]: Same comment as below about decorator.

File: examples/plotting/file/slider_pycalback.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+"""
```
Comment (bryevdv) [inline examples/plotting/file/slider_pycalback.py]: missing `l` in `pycallback` but I think a better name might be more like: `python_callback.py` thats the most important aspect to emphasize, the slider part is just a detail.

File: examples/plotting/file/slider_pycalback.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+@CustomJS
+def callback(source=source):
+    data = source.get('data')
+    A, B = amp.get('value'), offset.get('value')
+    k, phi = freq.get('value'), phase.get('value')
+    x, y = data['x'], data['y']
+    for i in range(len(x)):
+        y[i] = B + A * Math.sin(k * x[i] + phi)
+    source.trigger('change')
```
Comment (bryevdv) [inline examples/plotting/file/slider_pycalback.py]: Maybe not for now but I think perhaps this could be automated.

**Comment by almarklein** [comment]
Thanks @bryevdv. I agree with all comments, except perhaps the decorator. Having to do this seems a bit verbose to me. I don't like that you have to think of a new for a function that you're not actually using:

    def _my_callback(source=source, s=slider):
        ...
    callback = CustomJS(_my_callback)

What about a new decorator, e.g. `@make_custom_js` or `@custom_js_from_py`, or perhaps `@CustomJS.from_func` (or similar)?

**Comment by bryevdv** [comment]
Another option might be `transform` on the property that checks for python functions:

https://github.com/bokeh/bokeh/blob/master/bokeh/properties.py#L1069

I think perhaps `Callback` is to general for the property type here, actually.

**Comment by bryevdv** [comment]
@almarklein also note https://github.com/bokeh/bokeh/pull/3160 I am about to merge this, we can use it to mark dependencies. (In this case, `required` to use for python callbacks)

**Comment by almarklein** [comment]
> Another option might be transform on the property that checks for python functions:

I am not sure I follow. What would user code look like in that case?

**Comment by bryevdv** [comment]
`thing.callback = some_python_function`

**Comment by almarklein** [comment]
Right, I think I quite like that, it should also allow `Thing(callback=some_python_function)`.

**Comment by almarklein** [comment]
@bryevdv I tried to use `transform()` to allow `x.callback = py_func`, but that does not work, because callback properties are defined as `Instance(CallBack)`, and the in `validate()` of `Instance` is done before we get a chance to transform the value. Other thoughts?

We could folllow a similar implementation as it currently is, but write it up without a decorator:

    thing.callback = CustomJS(py_func)
    # or
    thing.callback = CustomJS.from_func(py_func)

**Comment by mattpap** [comment]
@almarklein, you should try `accepts(type, converter_fn)` on `callback` property (see `Legend.legends` as an example). The `converter_fn` part is trivial (`from_func()`), just the `type` is something that needs extra work, i.e. a new type of property. Actually, we have dummy property type `Function`, so you just need to implement type-checking there (`validate()` function).

**Comment by mattpap** [comment]
Commit d3c5600 is simply entitled "merge". What happened that there is no information about branches involved and what were the conflicts? I'm curious, because `git merge` does the right thing by default.

**Comment by almarklein** [comment]
@mattpap this needed a merge with master, because it had a small conflict due to the renaming of PlotObject to Model. Did I do something wrong there?

Thanks @mattpap that works.This means the `accepts()` trick will need to be applied everywhere where a `callback` property is used, which is not perfect, though it works.

But now I run into the problem that for e.g. the slider example, a new `CustomJS` object is created as we pass `Slider( ..., callback=py_func)` to each slider, and the example stops working. Therefore I think it makes much more sense to turn the python function into a `CustomJS` asap, before using it in sliders, etc.

I still like the idea of a decorator (though one with a better name), but writing `callback=CustomJS(py_callback)` or `callback=CustomJS.from_py_func(py_callback)` would work for me as well. Thoughts?

**Comment by bryevdv** [comment]
@almarklein I usually commit with `"merge master"` to be explicit in those cases.

Re: decorator, this still just seems like a very unusual an unexpected usage. My vote for what to demonstrate in the official docs/examples is `callback=CustomJS(py_callback)` (tho the decorator usage should still work if people wanted to use that)

**Comment by bryevdv** [comment]
As an aside, I find with a little usage now that I don't like `"CustomJS"` at all but I think that ship has sailed.

**Comment by mattpap** [comment]
> Did I do something wrong in the merge process?

Technically not, but I'm just curious about how you (actually Brian as well) do merges. I see those simplistic and hand-written merge commit messages often (not only in this project) and it's strange to me, because if you simply `git merge branch-name`, you will get the expected (by my self) commit message.

**Comment by mattpap** [comment]
> I find with a little usage now that I don't like "CustomJS" at all (...)

Especially it can do CoffeeScript as well.

**Comment by mattpap** [comment]
@almarklein, how much WIP this PR still is?

File: bokeh/models/callbacks.py [hunk 8]
```diff
@@ -24,6 +27,29 @@ class OpenURL(Callback):
...
+                                       'you need Flexx (pip install flexx)')
+            # Collect default values
+            default_values = func.__defaults__  # Python 2.6+
+            default_names = func.__code__.co_varnames[:len(default_values)]
+            kwargs['args'] = dict(zip(default_names, default_values))
+            # Get JS code, we could rip out the function def, or just
+            # call the function. We do the latter.
+            kwargs['code'] = py2js(func, 'cb') + 'cb(%s);\n' % ', '.join(default_names)
+
+        Callback.__init__(self, **kwargs)
```
Comment (mattpap) [inline bokeh/models/callbacks.py]: This code should force `lang` property to `javascript`.

Comment (mattpap) [inline bokeh/models/callbacks.py]: Going further, I think `kwargs` shouldn't be overridden silently when user specified `code` and/or `args`.

**Comment by bryevdv** [comment]
> because if you simply git merge branch-name, you will get the expected (by my self) commit message.

Only when there are not conflicts.

**Comment by mattpap** [comment]
> Only when there are not conflicts.

Strange, I created e.g. https://github.com/bokeh/bokeh/commit/56ed79bf75110c6a4782c6b894f817c7a852ef87 with `git merge origin/master`, which resulted in conflicts, which I resolved and then issued `git commit`, and the commit message contains all expected information. Anyway, this is off-topic, so we may want to discuss this in more appropriate place.

**Comment by fpliger** [comment]
> I still like the idea of a decorator (though one with a better name), but writing callback=CustomJS(py_callback) or callback=CustomJS.from_py_func(py_callback) would work for me as well. Thoughts?

Sorry to jump in late. I really dislike `callback=CustomJS(py_callback)`... I think `CustomJS` is ok (surely we could have chosen a better name but I agree this ship has sailed :) but the combination of it with `py_callback` is really awful. 

I like `callback=CustomJS.from_py_func(py_callback)` and maybe the best option we have.

**Comment by almarklein** [comment]
This is basically done. I just need a decision on the API. Should this PR wait for the conda package to be available?

On `git merge`, I did `git merge master`, solved the conflicts, then git told me to commit when resolved, so I did `git commit -m merge`.

**Comment by damianavila** [comment]
FYI, I have mane conda packages for flexx in all platforms with py34 and p35...

So, change needed:  modify `conda.recipe/meta.yaml` to selectively install flexx in py34 and py35

**Comment by almarklein** [comment]
Flexx is available in the bokeh channel: https://anaconda.org/bokeh/flexx/files

Travis is failing, but it seems unrelated to this PR, any ideas?

**Comment by mattpap** [comment]
@almarklein, it seems unrelated.

**Comment by mattpap** [comment]
This is LGTM from me, just lets see if CI doesn't fail to build using conda package.

**Comment by bryevdv** [comment]
Agreed

**Comment by jakirkham** [comment]
So, what was the final decision with the API? Is it this `CustomJS.from_py_func`?

**Comment by mattpap** [comment]
@jakirkham, it's `CustomJS.from_py_func`, as the safest approach, but I think a decorator or an implicit conversion will be added sooner than later.

**Comment by almarklein** [comment]
@mattpap With the current implementation you could very well do:

    @CustomJS.from_py_func
    def callback(..)
        ...

But most people seem to prefer the more explicit `callback = CustomJS.from_py_func(py_func)`, which is what we thus use in our examples.

**Comment by mattpap** [comment]
@almarklein, ah of course. But anyway, I think that we could add support for `callback = py_func` by using `accepts()`.

**Comment by bryevdv** [comment]
Yes, I definitely want to add some convenience I just want a chance to discuss the best interface.

**Comment by jakirkham** [comment]
Thanks @mattpap.

File: scripts/travis_install [hunk 9]
```diff
@@ -39,6 +39,9 @@ BK_PKG_NAME=${BK_PKG_NAME/nightly/$VERSION}
 
 conda install --force --yes $BK_PKG_NAME
 
+conda install --yes pip
```
Comment (damianavila) [inline scripts/travis_install]: now, you have flexx packages, you need to add flexx as the test dependency in the `conda.recipe/meta` file

Comment (damianavila) [inline scripts/travis_install]: sorry I commented before refreshing my page... nvm...

**Comment by mattpap** [comment]
> `scripts/travis_install: line 56: [: too many arguments`

@damianavila, this may be interesting to you.

Anyway, one test run fails (`python_tests`), because flexx is not installed.

**Comment by almarklein** [comment]
@mattpap I think the Travis fails are unrelared to this PR, but I don't understand why they're still there. I already merged in the latest master.

**Comment by almarklein** [comment]
green!

**Comment by bryevdv** [comment]
Failing on master because Flexx is not installed:

https://travis-ci.org/bokeh/bokeh/jobs/94046413#L415

@almarklein @damianavila why was this passing on the PR?

**Comment by bryevdv** [comment]
...caches

**Pull request merged**